dawn/src/dawn/AST/ASTStmt.h

```diff
@@ -111,6 +111,8 @@ class Stmt : public std::enable_shared_from_this<Stmt> {
     return dynamic_cast<DataType&>(*data_.get());
   }
 
+  bool hasData() const { return data_ != nullptr; }
```

Comment:
A statement has always a data object attached, also in SIR (it was envisioned to move SIR-specific structures such as `sir::VerticalRegion` into the respective stmt's data, but it never became a priority...). Therefore, a call to this will always return true.
It was envisioned to mirror the same pattern for data objects used in stmts also in expressions, but this was, again, never implemented (currently they are created on-the-fly in the first call to `getData()`, a horrible pattern ...).
From the usage below it seems that you introduced these `hasData()` to determine whether you are visiting SIR or IIR. As we start to specialize the unstruct. dim. checker's implementation for IIR-specific checks, I think a cleaner way to do this would be to introduce and propagate a boolean flag `isIIR` that would enable such checks.